lib/datadog/core/configuration/option.rb

```diff
@@ -94,6 +96,91 @@ def default_precedence?
 
         private
 
+        def coerce_env_variable(value)
+          case @definition.type
+          when :int
+            value.to_i
+          when :float
+            value.to_f
+          when :array
+            values = if value.include?(',')
+                       value.split(',')
+                     else
+                       value.split(' ') # rubocop:disable Style/RedundantArgument
+                     end
+
+            values.map! do |v|
+              v.gsub!(/\A[\s,]*|[\s,]*\Z/, '')
+
+              v.empty? ? nil : v
+            end
+
+            values.compact!
+            values
+          when :bool
+            string_value = value
+            string_value = string_value.downcase
+            string_value == 'true' || string_value == '1' # rubocop:disable Style/MultipleComparison
+          when :string, NilClass
+            value
+          else
+            raise ArgumentError, "The option #{@definition.name} is using an unknown type option `#{@definition.type}`"
+          end
+        end
+
+        def validate_type(value)
+          raise_error = false
+
+          valid_type = validate(@definition.type, value)
+
+          unless valid_type
+            raise_error = if @definition.type_options[:nil]
+                            !value.is_a?(NilClass)
+                          else
+                            true
+                          end
+          end
+
+          if raise_error
+            error_msg = if @definition.type_options[:nil]
+                          "The option #{@definition.name} support this type `#{@definition.type}` "\
+                                      "and `nil` but the value provided is #{value.class}"
+                        else
+                          "The option #{@definition.name} support this type `#{@definition.type}` "\
+                          "but the value provided is #{value.class}"
+                        end
+
+            raise ArgumentError, error_msg
+          end
+
+          value
+        end
+
+        def validate(type, value)
+          case type
+          when :string
+            value.is_a?(String)
+          when :int
+            value.is_a?(Integer)
+          when :float
+            value.is_a?(Float)
+          when :array
+            value.is_a?(Array)
+          when :hash
+            value.is_a?(Hash)
+          when :bool
+            value.is_a?(TrueClass) || value.is_a?(FalseClass)
+          when :proc
+            value.is_a?(Proc)
+          when :symbol
+            value.is_a?(Symbol)
+          when NilClass
+            true
+          else
+            raise ArgumentError, "The option #{@definition.name} is using an unknown type option `#{@definition.type}`"
```

Comment:
Should we have a specific exception (or exceptions) for such configuration validation errors?

I'm thinking of at least a few:

- `Configuration::Error`: generic error
- `Configuration::TypeError`: when type does not match
- `Configuration::ValueError`: when value is of the correct type but fails to pass a value check, esp. reusable for those `setter`-time checks
- `Configuration::ConsistencyError`: when e.g two settings are mutually exclusive but both are set in inconsistent ways
- etc...

Those would be automatic for simple types, but be especially reusable for those `setter`-time checks that can be much more rich.

Reply:
If there's enough repetition, the error-specific types seems fine.

If they are not very reusable today, then I think it's fine to leave them as they are today.

Reply:
To be honest, what's the value of the more specific types? My thinking is -- if you messed up with configuration, you need to fix it.

We don't expect customers to be rescuing any of the more specific types, so if you're not going to be rescuing them, is there a lot of value in having different types?